keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,566 @@
+
+Just like for existing Job patterns, workloads have to handle duplicates at the
+application level.
+
+[reference]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures
+
+## Design Details
+
+### JobSpec API
+
+The JobSpec gets the field `completionMode` to control whether the Job should	
+be treated as an Indexed Job.	
+
+```golang	
+type JobSpec struct {	
+  ...	
+  // CompletionMode specifies how Pod completions are tracked. It can be	
+  // `NonIndexed` (default) or `Indexed`.	
+  // `NonIndexed` means that each Pod completion is homologous to each other.	
+  // The Job is considered complete when there have been .spec.completions	
+  // successful completions.	
+  // `Indexed` means that each Pod completion needs to be tracked individually;	
+  // each Pods gets an associated completion index, which is available in the
+  // annotation	`batch.alpha.kubernetes.io/job-completion-index`.	
+  // The Job is considered complete when there is one successful Pod for each	
+  // index in the range 0 to (.spec.completions - 1).	
+  // When value is `Indexed`, .spec.completions must have a non-zero positive	
+  // value and `.spec.parallelism` must be less than or equal to 10^6.
+  //
+  // More completion modes can be added in the future. If a Job controller
+  // observes a mode that it doesn't recognize, it manages the Job as in
+  // `NonIndexed`.
+  CompletionMode string	
+}	
+```
+
+As the comment describes, when `.spec.completionMode = "Indexed"`:
+
+- The `.spec.completions` must be a non-zero positive value. This is to trigger
+  Jobmanagement strategy for *fixed completion count*. That is, `Indexed` mode
+  cannot	be used for work queue patterns.	
+- The `.spec.parallelism` must be up to `10^6`. This is to guarantee that we
+  can keep track of completions per-index in the Job status in the future.
+
+### Pod detail
+
+The Pod and PodSpec APIs don't get any new fields. However, Pods created for
+Indexed Jobs get the annotation `batch.alpha.kubernetes.io/job-completion-index`
+with a value equal to its completion index. The annotation is immutable.
+
+The annotation can be accessed through the downward API as a file or environment
+variable.
+
+For user convenience, the Job controller adds the completion index as an
+environment variable through the downward API. That is, the Job controller
+creates Pods like so:
+
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+    - name: test-container
+      env:
+        - name: JOB_COMPLETION_INDEX
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.annotations['batch.alpha.kubernetes.io/job-completion-index'] 
+```
+
+The Job controller doesn't add the environment variable if there is a name
+conflict with an existing environment variables in the Job's Pod template or if
+the user already specified another environment variable for the same annotation.
```

Comment:
Why? What's the problem with having the value referenced by two separate env vars?

Reply:
I think it makes sense to give the user control over the names of the variables. A binary could always have some behavior defined for JOB_COMPLETION_INDEX which is different from the one provided by k8s.

Reply:
I guess I wasn't clear. I wasn't asking to touch customer-defined variable.

I was just asking - why we don't want to provide JOB_COMPLETION_INDEX unconditionally (either with user-propagated-value if set by user or as above if not).

I don't see a problem if you have two env vars, say FOO and JOB_COMPLETION_INDEX with the same value set to the val of that annotation.

Reply:
I understood you correctly :)

What I'm saying is: what if a JOB_COMPLETION_INDEX variable causes an undesirable effect on the user binary?

I guess in this case they could override that environment variable with another value.

Reply:
I don't have a strong opinion about the name. But I'm still not convinced we should always inject the value if the user is already adding it.

Reply:
We should always inject, unless user explicitly overwrites that name.

The behavior should be consistent - if you job is indexed, var X (I don't have strong opinion about name) will always be set (in your behavior you have "the var may or may not be set" - that's super confusing).

Reply:
Sounds good. Removed second condition.